runtime/cudaq/dynamics/cudm_state.cpp

```diff
@@ -0,0 +1,252 @@
+std::vector<std::complex<double>> cudm_state::get_raw_data() const {
+  return rawData_;
+}
+
+std::vector<int64_t> cudm_state::get_hilbert_space_dims() const {
+  return hilbertSpaceDims_;
+}
+
+std::string cudm_state::dump() const {
+  if (!is_initialized()) {
+    throw std::runtime_error("State is not initialized.");
+  }
+
+  std::ostringstream oss;
+  oss << "State data: [";
+  for (size_t i = 0; i < rawData_.size(); i++) {
+    oss << rawData_[i];
+    if (i < rawData_.size() - 1) {
+      oss << ", ";
+    }
+  }
+  oss << "]";
+  return oss.str();
+}
+
+cudm_state cudm_state::to_density_matrix() const {
+  if (!is_initialized()) {
+    throw std::runtime_error("State is not initialized.");
+  }
+
+  if (is_density_matrix()) {
+    throw std::runtime_error("State is already a density matrix.");
+  }
+
+  std::vector<std::complex<double>> densityMatrix;
+  size_t vectorSize = calculate_state_vector_size(hilbertSpaceDims_);
+  size_t expectedDensityMatrixSize = vectorSize * vectorSize;
+  densityMatrix.resize(expectedDensityMatrixSize);
+
+  for (size_t i = 0; i < vectorSize; i++) {
+    for (size_t j = 0; j < vectorSize; j++) {
+      densityMatrix[i * vectorSize + j] = rawData_[i] * std::conj(rawData_[j]);
+    }
+  }
+
+  cudm_state densityMatrixState(densityMatrix);
+  densityMatrixState.init_state(hilbertSpaceDims_);
+  return densityMatrixState;
+}
+
+cudensitymatState_t cudm_state::get_impl() const {
+  if (!is_initialized()) {
+    throw std::runtime_error("State is not initialized.");
+  }
+  return state_;
+}
+
+void cudm_state::attach_storage() {
+  if (!state_) {
+    throw std::runtime_error("State is not initialized.");
+  }
+
+  if (rawData_.empty()) {
+    throw std::runtime_error("Raw data is empty. Cannot attach storage.");
+  }
+
+  // Retrieve the number of state components
+  int32_t numStateComponents;
+  HANDLE_CUDM_ERROR(
+      cudensitymatStateGetNumComponents(handle_, state_, &numStateComponents));
+
+  // Retrieve the storage size for each component
+  std::vector<size_t> componentBufferSizes(numStateComponents);
+  HANDLE_CUDM_ERROR(cudensitymatStateGetComponentStorageSize(
+      handle_, state_, numStateComponents, componentBufferSizes.data()));
+
+  // Validate that rawData_ has sufficient space for all components
+  size_t totalSize = 0;
+  for (size_t size : componentBufferSizes) {
+    totalSize += size;
+  }
+  if (rawData_.size() * sizeof(std::complex<double>) < totalSize) {
+    throw std::invalid_argument(
+        "Raw data size is insufficient to cover all components.");
+  }
+
+  // Attach storage for each component
+  std::vector<void *> componentBuffers(numStateComponents);
+  std::vector<std::complex<double> *> rawComponentData(numStateComponents);
+
+  size_t offset = 0;
+  for (int32_t i = 0; i < numStateComponents; i++) {
+    rawComponentData[i] =
+        reinterpret_cast<std::complex<double> *>(rawData_.data()) + offset;
+    componentBuffers[i] = static_cast<void *>(rawComponentData[i]);
```

Comment:
We're supposed to attached device memory here (`rawData_` is a host array).
I think we need to add a `gpuData_` member var and perform `cudaMalloc` + `cudaMemcpy`.